vnext/Desktop/Modules/WebSocketModule.cpp

```diff
@@ -148,27 +148,39 @@ std::shared_ptr<IWebSocketResource> WebSocketModule::GetOrCreateWebSocket(int64_
   {
     auto ws = IWebSocketResource::Make(std::move(url));
     auto weakInstance = this->getInstance();
-    ws->SetOnError([this, id, weakWs = weak_ptr<IWebSocketResource>(ws), instance = weakInstance.lock()](const IWebSocketResource::Error& err)
+    ws->SetOnError([this, id, weakInstance](const IWebSocketResource::Error& err)
     {
-      auto strongWs = weakWs.lock();
+      auto strongInstance = weakInstance.lock();
+      if (!strongInstance)
+        return;
+
       auto errorObj = dynamic::object("id", id)("message", err.Message);
       this->SendEvent("websocketFailed", std::move(errorObj));
     });
-    ws->SetOnConnect([this, id, weakWs = weak_ptr<IWebSocketResource>(ws), instance = weakInstance.lock()]()
+    ws->SetOnConnect([this, id, weakInstance]()
     {
-      auto strongWs = weakWs.lock();
+      auto strongInstance = weakInstance.lock();
+      if (!strongInstance)
+        return;
+
       auto args = dynamic::object("id", id);
       this->SendEvent("websocketOpen", std::move(args));
     });
-    ws->SetOnMessage([this, id, weakWs = weak_ptr<IWebSocketResource>(ws), instance = weakInstance.lock()](size_t length, const string& message)
+    ws->SetOnMessage([this, id, weakInstance](size_t length, const string& message)
     {
-      auto strongWs = weakWs.lock();
+      auto strongInstance = weakInstance.lock();
+      if (!strongInstance)
+        return;
+
       auto args = dynamic::object("id", id)("data", message)("type", "text");
       this->SendEvent("websocketMessage", std::move(args));
     });
-    ws->SetOnClose([this, id, weakWs = weak_ptr<IWebSocketResource>(ws), instance = weakInstance.lock()](IWebSocketResource::CloseCode code, const string& reason)
+    ws->SetOnClose([this, id, weakInstance](IWebSocketResource::CloseCode code, const string& reason)
     {
-      auto strongWs = weakWs.lock();
+      auto strongInstance = weakInstance.lock();
+      if (!strongInstance)
+        return;
+
       auto args = dynamic::object("id", id)("code", static_cast<uint16_t>(code))("reason", reason);
       this->SendEvent("websocketClosed", std::move(args));
```

Comment:
>this [](start = 6, length = 4)

we should be using strongInstance here

Reply:
And in other places

---
In reply to: [405146909](https://github.com/microsoft/react-native-windows/pull/4510#discussion_r405146909) [](ancestors = 405146909)

Reply:
As @NickGerleman pointed out, wouldn't this produce a reference cycle?

- `shared_ptr<Instance>` owns vector<unique_ptr<CxxModule>>, which will contain `WebSocketModule`.
- `WebSocketModule` owns `shared_ptr<IWebSocketResource>`.
- `WinRTWebSocketResource` owns the lambda (`std::function<void(...)>`).

Reply:
I think Nikolai meant to use it instead of `this` (correct me if I'm wrong). Reference cycles would happen if we capture the strong ptr. 

`this` and `strongInstance` point to separate things though. Not sure if raw `this` usage is safe here. Does the instance keep this alive? Capturing `this as a weak_ptr as well might make this easier to reason about.

Reply:
Oh, looks like Khalef said the same thing already.

Reply:
Thanks for explanation, please add detailed comments :)

---
In reply to: [405154747](https://github.com/microsoft/react-native-windows/pull/4510#discussion_r405154747) [](ancestors = 405154747)